protocols/request-response/src/cbor.rs

```diff
@@ -0,0 +1,180 @@
+// Copyright 2020 Parity Technologies (UK) Ltd.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a
+// copy of this software and associated documentation files (the "Software"),
+// to deal in the Software without restriction, including without limitation
+// the rights to use, copy, modify, merge, publish, distribute, sublicense,
+// and/or sell copies of the Software, and to permit persons to whom the
+// Software is furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+// OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+// DEALINGS IN THE SOFTWARE.
+
+use crate::{Config, ProtocolSupport, RequestId, ResponseChannel};
+use async_trait::async_trait;
+use futures::prelude::*;
+use futures::{AsyncRead, AsyncWrite};
+use libp2p_core::upgrade::{read_length_prefixed, write_length_prefixed};
+use libp2p_core::Multiaddr;
+use libp2p_identity::PeerId;
+use libp2p_swarm::{NetworkBehaviour, StreamProtocol};
+use serde::{de::DeserializeOwned, Serialize};
+use std::{io, marker::PhantomData};
+
+#[derive(Debug, Clone)]
+pub struct Codec<Req, Resp> {
```

Comment:
Make this private please.

Reply:
If we make this struct private, we get the error

```
error[E0446]: private type `json::Codec<Req, Resp>` in public interface
  --> protocols/request-response/src/json.rs:43:10
   |
34 | struct Codec<Req, Resp> {
   | ----------------------- `json::Codec<Req, Resp>` declared as private
...
43 | #[derive(NetworkBehaviour)]
   |          ^^^^^^^^^^^^^^^^ can't leak private type
```

Reply:
Somewhere here )))

```
impl<Req, Resp> libp2p_swarm::derive_prelude::NetworkBehaviour
    for Behaviour<Req, Resp>
    where
        Req: Send + Clone + Serialize + DeserializeOwned + 'static,
        Resp: Send + Clone + Serialize + DeserializeOwned + 'static,
        crate::Behaviour<
            Codec<Req, Resp>,
        >: libp2p_swarm::derive_prelude::NetworkBehaviour,
        OutEvent<
            Req,
            Resp,
        >: From<
            <crate::Behaviour<
                Codec<Req, Resp>,
            > as libp2p_swarm::derive_prelude::NetworkBehaviour>::ToSwarm,
        >,
    {
        
        type ConnectionHandler = libp2p_swarm::derive_prelude::THandler<
            crate::Behaviour<Codec<Req, Resp>>,
        >;
        
        type ToSwarm = OutEvent<Req, Resp>;
```

Reply:
It's reasonable that the codec impl can't be private 
due to it is used as a generic type in `Behaviour<TCodec>`.
I did it `pub(crate)`

Reply:
> I did it pub(crate)

Unfortunately, it doesn't work

Reply:
I pushed a commit. There is a trick you can apply: The lint for "private type in public API" only looks at the modifier for the type, not whether it is actually reachable. So you can put it into a private module to effectively hide it :)

Reply:
I did it for cbor, can you replicate it for the json module?

Reply:
Ok, you pay close attention to the Public API.